File: backend/app/schemas/problem.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.execution import SupportedLanguage


Difficulty = Literal["easy", "medium", "hard"]
ProblemStatus = Literal["not_started", "attempted", "solved"]
RuntimeShape = Literal["plain", "linked_list", "random_pointer_linked_list"]


class TestCaseCreate(BaseModel):
    input: Any
    expected_output: Any
    explanation: str = ""


class TestCaseDraft(BaseModel):
    input_json: str
    expected_output_json: str
    explanation: str = ""


class ProblemBase(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    slug: str = Field(min_length=1, max_length=255, pattern=r"^[a-z0-9-]+$")
    topic: str = Field(min_length=1, max_length=100)
    difficulty: Difficulty
    description: str = Field(min_length=1)
    runtime_shape: RuntimeShape = "plain"
    starter_code: str = Field(min_length=1)
    function_name: str = Field(min_length=1, max_length=120)
    visible_test_cases: list[TestCaseCreate]
    hidden_test_cases: list[TestCaseCreate] = []
    notes: str = ""
    similar_questions: str = ""

    @field_validator("slug")
    @classmethod
    def normalize_slug(cls, value: str) -> str:
        return value.strip().lower()

    @field_validator("topic", "title", "function_name")
    @classmethod
    def strip_text(cls, value: str) -> str:
        return value.strip()


class ProblemCreate(ProblemBase):
    pass


class ProblemGenerationRequest(BaseModel):
    problem_statement: str = Field(min_length=20)


class ProblemGenerationResponse(BaseModel):
    title: str
    slug: str
    topic: str
    difficulty: Difficulty
    description: str
    runtime_shape: RuntimeShape
    starter_code: str
    starter_codes: dict[SupportedLanguage, str]
    function_name: str
    visible_test_cases: list[TestCaseCreate]
    hidden_test_cases: list[TestCaseCreate]
    generation_warnings: list[str] = []


class ProblemSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    slug: str
    topic: str
    difficulty: Difficulty
    runtime_shape: RuntimeShape
    status: ProblemStatus
    updated_at: datetime


class ProblemDetail(ProblemSummary):
    description: str
    starter_code: str
    starter_codes: dict[SupportedLanguage, str]
    function_name: str
    visible_test_cases: list[TestCaseCreate]
    hidden_test_cases: list[TestCaseCreate]
    notes: str
    similar_questions: str
    created_at: datetime


class ProblemStatusUpdate(BaseModel):
    status: ProblemStatus


class ProblemWorkspaceUpdate(BaseModel):
    notes: str = ""
    similar_questions: str = ""


class ProblemListResponse(BaseModel):
    items: list[ProblemSummary]
    total: int

File: frontend/src/types/problem.ts
export type Difficulty = "easy" | "medium" | "hard";
export type ProblemStatus = "not_started" | "attempted" | "solved";
export type SupportedLanguage = "python" | "javascript" | "java" | "cpp";
export type RuntimeShape = "plain" | "linked_list" | "random_pointer_linked_list";

export interface TestCase {
  input: unknown;
  expected_output: unknown;
  explanation: string;
}

export interface ProblemSummary {
  id: number;
  title: string;
  slug: string;
  topic: string;
  difficulty: Difficulty;
  runtime_shape: RuntimeShape;
  status: ProblemStatus;
  updated_at: string;
}

export interface ProblemDetail extends ProblemSummary {
  description: string;
  starter_code: string;
  starter_codes: Record<SupportedLanguage, string>;
  function_name: string;
  visible_test_cases: TestCase[];
  hidden_test_cases: TestCase[];
  notes: string;
  similar_questions: string;
  created_at: string;
}

export interface ProblemListResponse {
  items: ProblemSummary[];
  total: number;
}

export interface ExecutionCaseResult {
  input: unknown;
  expected_output: unknown;
  actual_output: unknown | null;
  passed: boolean;
  stdout: string;
  error: string | null;
  explanation: string;
}

export interface ExecutionResponse {
  status: "success" | "error" | "timeout";
  mode: "run" | "submit";
  language: SupportedLanguage;
  passed_count: number;
  total_count: number;
  all_passed: boolean;
  results: ExecutionCaseResult[];
  error: string | null;
}

export interface SubmissionHistoryItem {
  id: number;
  mode: string;
  language: SupportedLanguage;
  code_snapshot: string;
  results: ExecutionCaseResult[];
  passed_count: number;
  total_count: number;
  all_passed: boolean;
  error: string | null;
  created_at: string;
}

export interface ProblemPayload {
  title: string;
  slug: string;
  topic: string;
  difficulty: Difficulty;
  description: string;
  runtime_shape: RuntimeShape;
  starter_code: string;
  function_name: string;
  visible_test_cases: TestCase[];
  hidden_test_cases: TestCase[];
  notes: string;
  similar_questions: string;
}

export interface ProblemGenerationResponse {
  title: string;
  slug: string;
  topic: string;
  difficulty: Difficulty;
  description: string;
  runtime_shape: RuntimeShape;
  starter_code: string;
  starter_codes: Record<SupportedLanguage, string>;
  function_name: string;
  visible_test_cases: TestCase[];
  hidden_test_cases: TestCase[];
  generation_warnings: string[];
}
